File: Leetcode 0127 Word Ladder [medium] [array]/Word Ladder [bfs].js
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
var ladderLength = function(beginWord, endWord, wordList) {
    let dic = new Set();
    for (let str of wordList) {
        dic.add(str);
    }
    
    let queue = [];
    queue.push(beginWord);
    let level = 0;
    
    while (queue.length !== 0) {
        let size = queue.length;
        level++;
        for (let i = 0; i < size; i++) {
            let current = queue.shift();
            if (current === endWord) {
                return level;
            }
            
            let tempStr = current.slice(0);
            for (let j = 0; j < current.length; j++) {
                for(let ch of "abcdefghijklmnopqrstuvwxyz") {
                    let tempStrArr = tempStr.split('');
                    tempStrArr[j] = ch;
                    tempStr = tempStrArr.join('');
                    
                    if (dic.has(tempStr)) {
                        queue.push(tempStr);
                        dic.delete(tempStr);
                    }
                }
                let tempStrArr = tempStr.split('');
                tempStrArr[j] = current[j];
                tempStr = tempStrArr.join('');
            }
        }
    }
    
    return 0;
};

File: Leetcode 0127 Word Ladder [medium] [array]/Word Ladder [bfs].java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        HashSet<String> dic = new HashSet<>();
        for (String str : wordList) {
            dic.add(str);
        }
        
        LinkedList<String> queue = new LinkedList<>();
        queue.offer(beginWord);
        int level = 0;
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            level++;
            for (int i = 0; i < size; i++) {
                String current = queue.poll();
                if (current.equals(endWord)) {
                    return level;
                }
                
                StringBuilder sb = new StringBuilder(current);
                
                for (int j = 0; j < current.length(); j++) {
                    for (char c = 'a'; c <= 'z'; c++) {
                        sb.setCharAt(j, c);
                        if (dic.contains(sb.toString())) {
                            queue.offer(sb.toString());
                            dic.remove(sb.toString());
                        }
                    }
                    sb.setCharAt(j, current.charAt(j));
                }
            }
        }
        
        return 0;
    }
}